genie-web/src/main/java/com/netflix/genie/web/services/impl/FileSystemAttachmentService.java

```diff
@@ -136,21 +147,83 @@ public void copyAll(final String id, final Path destination) throws IOException
      */
     @Override
     public void deleteAll(final String id) throws IOException {
-        final Path attachmentDir = this.attachmentDirectory.resolve(id);
+        this.deleteAttachments(id);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Set<URI> saveAttachments(final String jobId, final Set<Resource> attachments) throws IOException {
+        return this.writeAttachments(jobId, attachments);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void deleteAttachments(final String jobId) throws IOException {
+        final Path attachmentDir = this.attachmentDirectory.resolve(jobId);
         FileUtils.deleteDirectory(attachmentDir.toFile());
     }
 
-    private void writeAttachments(final String id, final Map<String, InputStream> attachments) throws IOException {
+    private Set<URI> writeAttachments(final String id, final Set<Resource> attachments) throws IOException {
         final Path requestDir = Files.createDirectories(this.attachmentDirectory.resolve(id));
+        final ImmutableSet.Builder<URI> uris = ImmutableSet.builder();
 
-        for (final Map.Entry<String, InputStream> entry : attachments.entrySet()) {
+        for (final Resource attachment : attachments) {
+            final String rawFilename = attachment.getFilename() == null
+                ? UUID.randomUUID().toString()
+                : attachment.getFilename();
             // Sanitize the filename
-            final Path fileName = Paths.get(entry.getKey()).getFileName();
+            final Path fileName = Paths.get(rawFilename).getFileName();
             final Path file = requestDir.resolve(fileName);
-            try (InputStream contents = entry.getValue()) {
+            try (InputStream contents = attachment.getInputStream()) {
                 final long byteCount = Files.copy(contents, file);
                 log.debug("Wrote {} bytes for attachment {} to {}", byteCount, fileName, file);
             }
+            uris.add(file.toUri());
+        }
+        return uris.build();
+    }
+
+    /**
+     * Temporary class for backwards compatibility till we delete all unused APIs.
```

Comment:
It's private so I figured it would be pretty obvious to remove once we blow away the API implementations as the IDE would show warnings for unused class.